Symbol checks

benbierens opened the pull request.

dryajov commented:
There is no need for handing this symbols in the wrapper.

You can use the 

```sh
  -t, --passC:OPTION        pass an option to the C compiler
  -l, --passL:OPTION        pass an option to the linker
```

options to the nim build which will forward them to the compiler. Or you can define {.passc: ... .} in the consuming application. Trying to handle this in the wrapper is almost always the wrong thing to do.

elcritch commented on leveldbstatic/raw.nim:
```diff
@@ -15,16 +16,23 @@ when defined(posix):
-{.passc: "-DHAVE_ZSTD=0".}
-{.passc: "-DHAVE_Zstd=0".}
+static:
+  proc doesCompile(cfile: string): int =
```
Unfortunately this will break if the system uses clang or a non-standard gcc, etc.

It'd probably be better to run and parse the cmake output and grep for these flags. I forget the command, but cmake does produce cached outputs of what flags are enabled. You can basically read it and grep for the results. However in doing that would really only make sense if `cmake` is re-run each time at compile time. That complicates things too. :)

It might make sense to just vendor and compile these libs as well. Basically go for the "fat" version of the lib! Though I think `fdatasync` might be OS specific.

Yet another option is to use something like `pkg-config` to check for the libraries.

Oh for `fullfsync` might consider using direct `when defined(linux) or defined(macosx): {.passc: ....}`. You could combine them with custom Nim defines as well, like `nim -d:leveldbFullSync` to allow user selection.

I'd probably go with vendored snappy, zstd, and the rest as enabled based on Nim defines and/or platform selects.

benbierens commented:
All right. I have removed the attempt at automatic checking and updated the documentation:
We default all the optimization options to false. (It works just fine and plenty fast enough this way.)
This allows users of the library to override these options and turn them on.

markspanbroek commented:
> It's totally possible to run cmake on every invocation, thus making the build more portable and allowing to enable flags at will. We're already doing this in leopard, please take a look - https://github.com/codex-storage/nim-leopard/blob/main/leopard/wrapper.nim

This would definitely work, but I'm personally not a big fan of this approach. It means that the nimble package has a dependency on cmake, which is not part of a standard nim installation. I usually try to avoid that dependency in nimble packages, for example in https://github.com/status-im/nim-ngtcp2

benbierens commented:
This latest commit removes the build script and prelude, and runs cmake on every nim build.
This adds a dependency to cmake, (git but that's always there for nimble), and increases build time by quite a bit.
The advantages I see to this approach is that pulling in a new version of leveldb becomes a little easier, and the compile-flags are automatically set to whatever's available on the system, while manual override is still possible.
I'm not sure if I like this approach better than the one used by Mark and Tanguy.

dryajov commented on leveldbstatic/include/port/port_config.h:
```diff
@@ -7,7 +7,7 @@
 // Define to 1 if you have a definition for fdatasync() in <unistd.h>.
 #if !defined(HAVE_FDATASYNC)
-#define HAVE_FDATASYNC 1
+#define HAVE_FDATASYNC 0
```
Why are we manually setting this?

markspanbroek replied:
This file no longer needs to be checked in, if we're always calling cmake, it will just generate this file for you.

dryajov commented:
> This would definitely work, but I'm personally not a big fan of this approach. It means that the nimble package has a dependency on cmake, which is not part of a standard nim installation. I usually try to avoid that dependency in nimble packages, for example in https://github.com/status-im/nim-ngtcp2

IMO, there is no reliable way around it if the wrapped library relies on `cmake`. When you generate the sources, it will only reflect the arch, os and third party deps available on the host as opposed to whats available on the target environment, which is definitely not what you want most of the time. Since we're interoping with a third party library written in a different language, it is unreasonable to constrain the toolset to only what nim offers.

dryajov reviewed (approved):
LGTM
